src/main/java/net/minecraftforge/fluids/FluidType.java

```diff
@@ -0,0 +1,1244 @@
+     * to {@code fluid_type.<modid>.<registry_name>}.
+     *
+     * @return the identifier representing the name of the fluid type
+     */
+    public String getDescriptionId()
+    {
+        if (this.descriptionId == null)
+            this.descriptionId = Util.makeDescriptionId("fluid_type", ForgeRegistries.FLUID_TYPES.get().getKey(this));
+        return this.descriptionId;
+    }
+
+    /**
+     * Returns how much the velocity of the fluid should be scaled by.
+     *
+     * @return a scalar to multiply to the fluid velocity
+     */
+    public double motionScale()
+    {
+        return this.motionScale;
+    }
+
+    /**
+     * Returns whether the fluid can push.
+     *
+     * @return {@code true} if the fluid can push, {@code false} otherwise
+     */
+    public boolean canPushEntity()
+    {
+        return this.canPushEntity;
+    }
+
+    /**
+     * Returns whether the fluid can be swum in.
+     *
+     * @return {@code true} if the fluid can be swum in, {@code false} otherwise
+     */
+    public boolean canSwim()
+    {
+        return this.canSwim;
+    }
+
+    /**
+     * Returns whether the fluid can drown something.
+     *
+     * @return {@code true} if the fluid can drown something, {@code false} otherwise
+     */
+    public boolean canDrown()
+    {
+        return this.canDrown;
+    }
+
+    /**
+     * Returns how much the fluid should scale the damage done when hitting
+     * the ground per tick.
+     *
+     * @return a scalar to multiply to the fall damage
+     */
+    public float getFallDistanceModifier()
+    {
+        return this.fallDistanceModifier;
+    }
+
+    /**
+     * Returns whether the fluid can extinguish.
+     *
+     * @return {@code true} if the fluid can extinguish, {@code false} otherwise
+     */
+    public boolean canExtinguish()
+    {
+        return this.canExtinguish;
+    }
+
+    /**
+     * Returns whether the fluid supports boating.
+     *
+     * @return {@code true} if the fluid supports boating, {@code false} otherwise
+     */
+    public boolean supportsBoating()
+    {
+        return this.supportsBoating;
+    }
+
+    /**
+     * Gets the path type of this fluid.
+     *
+     * @return the path type of this fluid
+     */
+    @Nullable
+    public BlockPathTypes getBlockPathType()
+    {
+        return this.pathType;
+    }
+
+    /**
+     * Gets the path type of the adjacent fluid. Path types with a negative
+     * malus are not traversable. Pathfinding will favor paths consisting of
+     * a lower malus.
+     *
+     * @return the path type of this fluid
+     */
+    @Nullable
+    public BlockPathTypes getAdjacentBlockPathType()
+    {
+        return this.adjacentPathType;
+    }
+
+    /**
+     * Returns whether the fluid can create a source.
+     *
+     * @return {@code true} if the fluid can create a source, {@code false} otherwise
+     */
+    public boolean canConvertToSource()
```

Comment:
I'm not convinced by the presence of these no-argument accessors. This looks like a pattern seen in vanilla, where a no-argument method (say, `Block.getFriction`), is basically patched out in favor of a Forge added one, with additional context (say, `IForgeBlock.getFriction`). However, this entirely relies on the fact that `Block.getFriction` is essentially deprecated, with instructions to never call this method (because blocks may be expected to override the `IForgeBlock` method, and will return inconsistent results).

In a `FluidType`, if a mod overrides `canConvertToSource(FluidState, LevelReader, BlockPos)`, and they (for example) only allow you to convert to sources in a certain dimension. The return value of `canConvertToSource()` is going to be just plain wrong. And yet there's no indication that this is a method that *should not be called*

It appears to me that `FluidType` is apparently designed to allow mods to override these methods and use the additional context they provide, except they *also* need to be able to function without any additional context, making said additional context a bit meaningless.

---

**Suggestion**:

Unless there's a vanilla call path that is documented (i.e. in `FluidType.getLightLevel`), then there should not be no-arg accessors for these properties. The only ones should be through the accessors with additional arguments, to actually support mods that override the latter.

Reply:
That's a fair enough point. Though, the one you are suggesting with `#canConvertToSource` needs to be dealt with in `FlowingFluid` unless we are completing deprecating the context, which I'm not sure how much I should do in vanilla. For right now, I'll assume that contexts which have been replaced will have their accessors removed as well.

Reply:
I don't see a way in which the methods can be effectively overridden from an API perspective, if there are still consistent enough call paths - either used by vanilla or from mods - that call no-arg methods. That just seems to be asking for inconsistent behavior depending on where the method is invoked from. So I'd say either:

- Remove the additional context, because it *will* be called through places that don't have it - so whatever is doing there needs to be able to handle said context. or,
- Fully patch in the additional context, and remove all (or "essentially" all) vanilla call paths for methods without context. (The path Forge typically takes with cases i.e. `IForgeBlock.getFriction()`). In this case also throwing `@Deprecated /** @deprecated call FluidType#canConvertToSource(...additional context...)` on the vanilla method `FlowingFluid#convertToSource()` would also be useful to indicate they shouldn't be used.